ext/BesselsEnzymeCoreExt.jl

```diff
@@ -0,0 +1,39 @@
+module BesselsEnzymeCoreExt
+
+  # TODO (cg 2023/05/08 10:02): Compat of any kind. 
+
+  using Bessels, EnzymeCore
+  using EnzymeCore.EnzymeRules
+  using Bessels.Math
+
+  # A manual method that takes an NTuple of partial sum terms and checks if it is
+  # exactly converged before computing the Levin sequence transformation. If it is
+  # exactly converged, the sequence transformation will create a divide-by-zero
+  # problem. See
+  #
+  # https://github.com/JuliaMath/Bessels.jl/issues/96 
+  #
+  # and links with for discussion.
+  #
+  # TODO (cg 2023/05/08 10:00): I'm not entirely sure how best to "generalize"
+  # this to cases like a return type of DuplicatedNoNeed, or something being a
+  # `Enzyme.Const`. These shouldn't in principle affect the "point" of this
+  # function (which is just to check for convergence before applying a
+  # function), but on its face this approach would mean I need a lot of
+  # hand-written extra methods. I have an open issue on the Enzyme.jl repo at
+  #
+  # https://github.com/EnzymeAD/Enzyme.jl/issues/786
+  #
+  # that gets at this problem a bit. But it's a weird request and I'm sure Billy
+  # has a lot of asks on his time.
+  function EnzymeRules.forward(func::Const{typeof(levin_transform)}, 
+                               ::Type{<:Duplicated}, 
+                               s::Duplicated,
+                               w::Duplicated)
+    (sv, dv, N) = (s.val, s.dval, length(s.val))
+    ls  = (sv[N-1] == sv[N]) ? sv[N] : levin_transform(sv, w.val)
```

Comment:
Is there a reason why this is sv[N]and not an explicit check if the weights are zero? 

I'm kinda wondering if the generic levin_transform case could be improved. Like if we feed it a weight vector of zero instead of just returning the partial sum it will return NaN instead of the right value. That seems to be a generic error with the function in general?

Reply:
I guess the issue is that the weights are zero in the half integer case but the derivative of the weights are non zero

Reply:
Just a general question.

So we have explicit checks on the last and second to last term in the sequence. If they are equal we return the last term in the sequence because we are operating under the assumption that the sequence has already converged. There is then no reason to do the sequence transformation on a sequence that has already converged.

There still is this weird thing that `sv[N-1] == sv[N]` is true but the weights are non-zero. This will happen for non half integer orders when the sequence converges fast after a couple of terms. This is pretty harmless I think but it is burying some subtlety with what is going on.

Now if `sv[N-1] == sv[N]` then we perform the sequence transformation on the sequence. Now my next question is does Enzyme try to differentiate this? Or rather how fast is this? The levin_transform itself should be inactive since we just need to feed the derivative sequence itself into the transform.

Also, are we still seeing an allocation with this?

Reply:
I guess my general question is should the levin_transform have a check within the function itself. Something like `all(i -> !isinf(i), w)` so if the weights are infinity we can then return the previous term of the sequence as we know it's converged? We can check this within one pass as we define the sequence and weights before the transform itself..... this might just be due to my misunderstanding of what the Enzyme is actually doing for us?

Reply:
A couple responses:

> Like if we feed it a weight vector of zero instead of just returning the partial sum it will return NaN instead of the 
> right value. That seems to be a generic error with the function in general?

I suppose, but considering that this is an internal function that we are not inviting users to directly access, I personally don't see the point of trying to get something like that perfect.

> Now if sv[N-1] == sv[N] then we perform the sequence transformation on the sequence

Do you mean "Now if sv[N-1] != sv[N] [...]"? Because if `sv[N-1] == sv[N]`, that is precisely when we do not do the transform on `sv`.

> The `levin_transform` itself should be inactive 

`Enzyme`'s `inactive` will mean the derivative is zero, so that concept isn't actually helpful here.

In general, here is how I would phrase the point of this method: `s.val` gives the tuple of partial series terms, and `s.dval` gives the tuple of _derivative_ partial series terms. In the case where `isinteger(v +1/2)`, `s.val` exactly converges in a few terms, but `s.dval` does **not**. So the code at present checks for each of those things to have converged on their own and transforms/accelerates whichever of the two have not converged. 

> I guess my general question is should the levin_transform have a check within the function itself.

Remember that this is effectively where we started: the function had an early return that checked for convergence. But the whole reason that that doesn't work is because you will have convergence of the series (`s.val`) when `isinteger(v+1/2)`, and so you'll return early, but that is incorrect because `s.dval` has **not** converged.

Reply:
This is very helpful!! Ignoring derivatives just for now. I was worried about the scalar case when we call something like...
```julia
julia> Bessels.BesselFunctions.besselkx_levin(0.5, 10.0, Val(17))
NaN
```
Which because we are not going to have explicit checks anymore (so want functions to work for any order input) that will give us an incorrect value. 

Now that is possible to fix pretty easily by changing within the levin_transform the line...
```julia
@nexprs $N i -> a_{i} = isinf(w[i]) ? (return s[i]) : Vec{2, T}((s[i] * w[i], w[i]))

# works now
julia> Bessels.BesselFunctions.besselkx_levin(0.5, 10.0, Val(17))
0.3963327297606011
```
Now that check should come after your top level check (so the rule is good I think) but the way it is currently set up if you call besselkx_levin at half integers it will still return NaN which we need to check. I think that should still work with the way you set it up though?

Reply:
So we use the same rule but we don't need the top level check anymore we just do the levin transform separately. But now the levin_transform will terminate early (won't give NaNs) because we make the check within the function instead of within the rule. The rule is to just do the levin transforms separately for both the series.

My comment about the inactivity was erroneous for sure. The rule is actually doing the levin transform separately I was just worried it was trying to do some quotient derivative within the levin transform which is not needed. My mistake.

Reply:
Ok perfect I'm hoping this will work! We will probably have a few gotchas that I was using the old versions that did 1 / weights and this feeds in just the weights then inverts within the levin_transform. I think once that is squared away this should work.

Again, your rule is great and needed (I finally grasped it) but need to also make sure we have the scalar case working (which was a mistake in my initial implementation).